### ISSUE TITLE
Performance issue when working on Irmin

### ISSUE THREAD

**Issue by voodoos** [ocamllabs/vscode-ocaml-platform#954]
Using `ocaml-lsp-server` 1.11.5 and the latest vscode extension the language server is mostly unusable on some Irmin's files.

For example in file `src/pack/mem/irmin_pack_mem.ml` most action around line 171 take more than 10 seconds...
```
[Trace - 2:43:51 PM] Received response 'textDocument/hover - (18)' in 60162ms.
[Trace - 2:44:21 PM] Received response 'textDocument/hover - (20)' in 87246ms.
[Trace - 2:43:14 PM] Received response 'textDocument/codeLens - (15)' in 29389ms.
[Trace - 2:44:36 PM] Received response 'textDocument/codeAction - (22)' in 99948ms.
```

We reproduced that on two laptops running macos.

Running the same queries directly with Merlin from the command line takes less than a second.

Surprisingly we were not able to reproduce the issue on VIM with `neovim lsp` and the same latest `ocaml-lsp-server` so my guess is that it is related to the vscode extension.

In fact it just got worse after restarting vscode: when the OCaml Platform is enabled the extension host closes repeatedly. The crash only happen with the Irmin workspace and I do have gitlens activated (without any issue until this).
(probably related to #849)

cc @zineb-ada

**Comment by voodoos** [comment]
Found the log in the console:

<img width="1227" alt="image" src="https://user-images.githubusercontent.com/5031221/168853033-5d46536f-06d1-4249-8e8b-0083bfbdba2b.png">

**Comment by ulugbekna** [comment]
Some notes: 

- I was able to reproduce both with master (9762f8e66e3a08cda858f94fb1a2105d294e1ab1) and 1.10.6
- I noticed that hover requests take ages to complete and trigger ocamllsp to consume >90% CPU 

<img width="966" alt="image" src="https://user-images.githubusercontent.com/16353531/169506319-e6213db4-969f-4dc8-9ea1-963beaa6957d.png">

- I collected `sample` when ocamllsp was consuming much CPU [here]()

- I have a hypothesis as to why vscode client is having problems only: vscode may be smarter than other clients when it comes to sending `cancelRequest` notifications, which may be causing this problem. It may also be the case that the problem comes from how we run ppxes.

**Comment by ulugbekna** [comment]
I think I can reproduce this in `ocaml-lsp-server/vendor/merlin/src/ocaml/parsing/parstree.mli` hovering over say `module_declaration` use or trying to jump to its definition. 

macos `sample` shows this: 

```
    8564 Thread_174668: Tick0
      8564 thread_start  (in libsystem_pthread.dylib) + 15  [0x7fff2033f443]
        8564 _pthread_start  (in libsystem_pthread.dylib) + 224  [0x7fff203438fc]
          8561 caml_thread_tick  (in ocamllsp) + 127  [0x100c59eef]  st_pthreads.h:293
          ! 8561 __select  (in libsystem_kernel.dylib) + 10  [0x7fff20316656]
          2 caml_thread_tick  (in ocamllsp) + 127  [0x100c59eef]  st_pthreads.h:295
          1 caml_thread_tick  (in ocamllsp) + 151  [0x100c59f07]  st_pthreads.h:292
            1 pthread_getspecific  (in libsystem_pthread.dylib) + 0  [0x7fff2033f4ad]
```

do I understand correctly that there are too many __select calls?

**Comment by rgrinberg** [comment]
Looking at the number and declaring too many is a bit subjective. There's a more reliable way to determine if there's indeed too many: a watcher is being woken up at every loop iteration without it being read. Instrument `lev` to maintain some info about the lifetime of watches.

For example, keep a global table of all registered watches. Add a watch to the table on `start`, increment the count on every time the callback is called, and eliminate the watch after every destroy. You can keep a backtrace on `start` for every watch to have a rough idea where it's coming from. After this experiment, you will have a good idea the watch that is waking up the code spuriously.

**Comment by rgrinberg** [comment]
@voodoos is this still an issue?

**Comment by voodoos** [comment]
Just reproduced, it is still painfully slow:

Calling Merlin in single or server mode is equally slow:
(cc @pitag-ha)
```
❯ ocamlmerlin single type-enclosing -position 146:39 -filename src/irmin-pack/mem/irmin_pack_mem.ml <src/irmin-pack/mem/irmin_pack_mem.ml | jq '.timing'
{
  "clock": 11399,
  "cpu": 11321,
  "query": 11153,
  "pp": 0,
  "reader": 1,
  "ppx": 23,
  "typer": 144,
  "error": 0
}
```

I find surprising the fact typing is so fast but the answer so slow for the `type-enclosing` query. There is no `cmt` loading or shape reduction involved in that query...

**Comment by rgrinberg** [comment]
>  find surprising the fact typing is so fast but the answer so slow for the type-enclosing query

Agreed. That looks quite puzzling. It's worth running this under `perf`.

>  There is no cmt loading or shape reduction involved in that query...

What about cmi loading? How many cmi's are there and how big are they?

Do you want to keep the issue open here? I don't see much that can be done on the vscode side.

**Comment by voodoos** [comment]
Cmi loading is done during the typing phase so it should count there.
You can close if you want, I bumped a related issue in Merlin's tracker.
